File: JAVA/OOP/lab6/modules/set.js
/**
 * Set module.
 * @module set
 */
import Device from './device.js';

/**
 * Checks if argument is integer.
 * @param {any} num
 * @returns {Boolean}
 */
function isNum(num) {
  return Number.isInteger(num);
}

/**
 * Class representing a set of powered devices .
 */
export default class Set {
  static setOfDevices = [];

  /**
   * Returns a total price of the bouquet.
   * @returns {Number} total price
   */
  static getTotalPower() {
    return Set.setOfDevices.reduce((acc, {
      power
    }) => acc + parseFloat(power), 0);
  }

  /**
   * Returns sorted set per power
   * @returns {Array} sorted array.
   */
  static sortPower() {
    return Set.setOfDevices.sort((next, prev) => parseInt(prev.power) - parseInt(next.power));
  }

  /**
   * Filters the devices by the specified EM radiation interval.
   * @param {Number} start beggining of the range.
   * @param {Number} stop end of the range.
   * @returns {Array} filtered array.
   */
  static sortRad(start, stop) {
    if (start < 0 || stop < 0 || start >= stop || !isNum(start) || !isNum(stop)) {
      console.error('Invalid arguments');
      return new Error('Invalid arguments');
    }
    return Set.setOfDevices
      .filter(({
        rad
      }) => rad >= start && rad <= stop);
  }
}

File: JAVA/OOP/lab6/src/modules/Set.java
package modules;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.stream.Collectors;

public class Set {

    static ArrayList<Device> setOfDevices = new ArrayList<>(Collections.emptyList());

    public static float getTotalPower() {
        return Set.setOfDevices.stream().map(device -> device.power).reduce((float) 0, Float::sum);
    }

    public static String getSortedByPower() {
        return Set.setOfDevices.stream().sorted(Comparator.comparingDouble(prev -> prev.power))
                .map(device -> String.format("%s, consumes %.2f watts", device.getType(), device.power))
                .collect(Collectors.joining("; "));
    }

    public static String getFromRadRange(float start, float end) {
        if (start < 0 || end < 0 || start >= end) {
            throw new Error("Invalid values of borders!");
        }
        String list = Set.setOfDevices.stream().filter(device -> device.radiation >= start && device.radiation <= end)
                .map(device -> String.format("%s, emits %.2f Hz", device.getType(), device.radiation))
                .collect(Collectors.joining("; "));
        return String.format("Devices which emitting from %s Hz to %s Hz: \n", start, end) + list;
    }

}